File: frontend/src/types/incident.ts
export type IncidentStatus =
  | 'new'
  | 'assigned'
  | 'in_progress'
  | 'pending'
  | 'resolved'
  | 'closed'
  | 'cancelled'

export type IncidentPriority =
  | 'p1_critical'
  | 'p2_high'
  | 'p3_medium'
  | 'p4_low'

export interface Incident {
  id: string
  title: string
  description: string
  status: IncidentStatus
  priority: IncidentPriority
  category: string | null
  subcategory: string | null
  reporter_id: string
  assignee_id: string | null
  sla_due_at: string | null
  resolved_at: string | null
  closed_at: string | null
  created_at: string
  updated_at: string
}

export interface StatusLog {
  id: string
  incident_id: string
  from_status: IncidentStatus | null
  to_status: IncidentStatus
  changed_by_id: string
  comment: string | null
  created_at: string
}

export interface IncidentDetail extends Incident {
  status_logs: StatusLog[]
}

export interface SLAStatus {
  incident_id: string
  sla_due_at: string | null
  is_overdue: boolean
  remaining_minutes: number | null
}

export interface CreateIncidentRequest {
  title: string
  description: string
  priority?: IncidentPriority
  category?: string
  subcategory?: string
  assignee_id?: string
}

export interface UpdateIncidentRequest {
  title?: string
  description?: string
  priority?: IncidentPriority
  category?: string
  subcategory?: string
}

export interface TransitionRequest {
  to_status: IncidentStatus
  comment?: string
  assignee_id?: string
}

export interface AssignRequest {
  assignee_id: string
  comment?: string
}

export const STATUS_LABELS: Record<IncidentStatus, string> = {
  new: '新規',
  assigned: '割り当て済み',
  in_progress: '対応中',
  pending: '保留中',
  resolved: '解決済み',
  closed: 'クローズ',
  cancelled: 'キャンセル',
}

export const PRIORITY_LABELS: Record<IncidentPriority, string> = {
  p1_critical: 'P1 緊急',
  p2_high: 'P2 高',
  p3_medium: 'P3 中',
  p4_low: 'P4 低',
}

export const STATUS_COLORS: Record<IncidentStatus, string> = {
  new: 'bg-blue-100 text-blue-800',
  assigned: 'bg-purple-100 text-purple-800',
  in_progress: 'bg-yellow-100 text-yellow-800',
  pending: 'bg-orange-100 text-orange-800',
  resolved: 'bg-green-100 text-green-800',
  closed: 'bg-gray-100 text-gray-800',
  cancelled: 'bg-red-100 text-red-800',
}

export const PRIORITY_COLORS: Record<IncidentPriority, string> = {
  p1_critical: 'bg-red-100 text-red-800 font-bold',
  p2_high: 'bg-orange-100 text-orange-800',
  p3_medium: 'bg-yellow-100 text-yellow-800',
  p4_low: 'bg-green-100 text-green-800',
}

File: backend/app/models/incident.py
import enum
import uuid
from datetime import datetime
from typing import TYPE_CHECKING, Optional

from sqlalchemy import Boolean, Enum as SAEnum, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID

from app.core.database import Base

if TYPE_CHECKING:
    from app.models.user import User


class IncidentStatus(str, enum.Enum):
    NEW = "new"
    ASSIGNED = "assigned"
    IN_PROGRESS = "in_progress"
    PENDING = "pending"
    RESOLVED = "resolved"
    CLOSED = "closed"
    CANCELLED = "cancelled"


class IncidentPriority(str, enum.Enum):
    P1_CRITICAL = "p1_critical"
    P2_HIGH = "p2_high"
    P3_MEDIUM = "p3_medium"
    P4_LOW = "p4_low"


class Incident(Base):
    __tablename__ = "incidents"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    status: Mapped[IncidentStatus] = mapped_column(
        SAEnum(IncidentStatus, name="incident_status"),
        nullable=False,
        default=IncidentStatus.NEW,
    )
    priority: Mapped[IncidentPriority] = mapped_column(
        SAEnum(IncidentPriority, name="incident_priority"),
        nullable=False,
        default=IncidentPriority.P3_MEDIUM,
    )
    category: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    subcategory: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    reporter_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    assignee_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    sla_due_at: Mapped[Optional[datetime]] = mapped_column(nullable=True)
    resolved_at: Mapped[Optional[datetime]] = mapped_column(nullable=True)
    closed_at: Mapped[Optional[datetime]] = mapped_column(nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), onupdate=func.now(), nullable=False
    )

    reporter: Mapped["User"] = relationship("User", foreign_keys=[reporter_id])
    assignee: Mapped[Optional["User"]] = relationship(
        "User", foreign_keys=[assignee_id]
    )
    status_logs: Mapped[list["IncidentStatusLog"]] = relationship(
        "IncidentStatusLog",
        back_populates="incident",
        order_by="IncidentStatusLog.created_at",
    )


class IncidentStatusLog(Base):
    __tablename__ = "incident_status_logs"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    incident_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("incidents.id"), nullable=False, index=True
    )
    from_status: Mapped[Optional[IncidentStatus]] = mapped_column(
        SAEnum(IncidentStatus, name="incident_status"), nullable=True
    )
    to_status: Mapped[IncidentStatus] = mapped_column(
        SAEnum(IncidentStatus, name="incident_status"), nullable=False
    )
    changed_by_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    comment: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), nullable=False
    )

    incident: Mapped["Incident"] = relationship(
        "Incident", back_populates="status_logs"
    )
    changed_by: Mapped["User"] = relationship("User")


class SLAPolicy(Base):
    __tablename__ = "sla_policies"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    priority: Mapped[IncidentPriority] = mapped_column(
        SAEnum(IncidentPriority, name="incident_priority"),
        unique=True,
        nullable=False,
    )
    response_time_minutes: Mapped[int] = mapped_column(Integer, nullable=False)
    resolution_time_minutes: Mapped[int] = mapped_column(Integer, nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), onupdate=func.now(), nullable=False
    )
